iceoryx_posh/doc/chunk_header.md

```diff
@@ -0,0 +1,218 @@
+
+1. No custom header and payload alignment doesn't exceed the `ChunkHeader` alignment
+
+```
+chunkSize = sizeof(chunkHeader) + payloadSize;
+```
+
+2. No custom header and payload alignment exceeds the `ChunkHeader` alignment
+
+Worst case scenario is when a part of the `ChunkHeader` crosses the payload alignment boundary, so that the payload must be aligned to the next boundary.
+```
+sizeLeftOfThePayloandAlignmentBoundary = sizeof(chunkHeader) - alignof(chunkHeader);
+chunkSize = sizeLeftOfThePayloandAlignmentBoundary + payloadAlignment + payloadSize;
+```
+
+3. Custom header is used
+
+Similar to case 2, but in this case it is the `back-offset` which might cross the payload alignment boundary.
+```
+headerSize = sizeof(chunkHeader) + sizeof(customHeader)
+sizeLeftOfThePayloandAlignmentBoundary = align(headerSize, alignof(payloadOffset));
+maxAlignment = max(alignof(payloadOffset), payloadAlignment);
+chunkSize = sizeLeftOfThePayloandAlignmentBoundary + maxAlignment + payloadSize;
+```
+
+### Accessing Chunk Header Extension
+
+The `ChunkHeader` has a template method to get custom header. There is a risk to use this wrong by accident, but there is currently no better solution for this.
+
+Since the custom header is always adjacent to the `ChunkHeader`, the formula to get the custom header is:
+```
+customHeader = addressOf(chunkHeader) + sizeof(chunkHeader);
+```
+
+### ChunkHeader Methods
+
+- `void* payload()` returns a pointer to the payload
+- `template <typename T> T* customHeader()` returns a pointer to the custom header
+- `static ChunkHeader* fromPayload(const void* const payload)` returns a pointer to the `ChunkHeader` associated to the payload
+
+### Integration Into Publisher/Subscriber API
+
+The `Publisher` has additional template parameters for the custom header and payload alignment. Alternatively this could also be done with the `allocate` method, but that increases the risk of using it wrong by accident.
+
+Additionally, to fully integrate the custom header into the regular control flow, a pointer to a `ChunkHeaderHook` can be passed to the `Publisher` c'tor.
+
+The `ChunkHeaderHook` is not in the shared memory and has the following virtual methods:
+- `virtual void allocateHook(ChunkHeader& chunkHeader)` can be used to initialize the custom header
+- `virtual void deliveryHook(ChunkHeader& chunkHeader)` can be used to set members of the custom header, e.g. a timestamp
+
+Furthermore, the `Publisher` and `Subscriber` have access to the `ChunkHeader` and can use the `customHeader()` method to gain access to the custom header.
+
+### Pitfalls & Testing
+
+- when the payload is adjacent to the `ChunkHeader`, it must be ensured that `m_payloadOffset` overlaps with the `back-offset`, which is `sizeof(m_payloadOffset)` in front of the payload
+- to simplify calculation, it is assumed that the alignment of the custom header doesn't exceed the alignment of the `ChunkHeader`. This has to be enforced with an `assert`
```

Comment:
I think it is a complex topic and I suggest solving this issue right away otherwise one of us as to get into the offset and alignment calculation next year and figure everything out again. 
Where would be the challenge? Could we get some `maxHeaderAlignment = std::max(alignof(ChunkHeader), customHeaderAlignment)` or do I miss something?

Reply:
if the alignment of the custom header is larger than the `ChunkHeader` we cannot assume that the position of the custom header is at `addressof(chunkHeader) + sizeof(chunkHeader)` which means we need to also store the custom header offset. I cannot think of a scenario where you want a special alignment for the custom header and also no scenario where it is necessary that the alignment should be larger than 8 bytes